ts/Series/Treegraph/TreegraphLayout.ts

```diff
@@ -0,0 +1,367 @@
+        if (node.isLeaf()) {
+            leftSibling = node.getLeftSibling();
+            if (leftSibling) {
+                node.preX = leftSibling.preX + siblingDistance;
+                node.mod = node.preX;
+            } else {
+                node.preX = 0;
+            }
+        } else {
+            // if the node has children perform the recursive first walk for its
+            // children, and then calculate its shift in the apportion function
+            // (most crucial part part ofthe algorythm)
+            let defaultAncestor = node.getLeftMostChild() as TreegraphNode;
+
+            node.linksFrom.forEach(function (link): void {
+                treeLayout.firstWalk(link.toNode as any);
+                defaultAncestor = treeLayout.apportion(
+                    link.toNode,
+                    defaultAncestor
+                );
+            });
+            treeLayout.executeShifts(node);
+
+            let leftChild = node.getLeftMostChild() as TreegraphNode;
+            let rightChild = node.getRightMostChild() as TreegraphNode;
+            // set the position of the parent as a middle point of its children
+            // and move it by the value of the leftSibling (if it exists)
+            let midPoint = (leftChild.preX + rightChild.preX) / 2;
+            leftSibling = node.getLeftSibling();
+
+            if (leftSibling) {
+                node.preX = leftSibling.preX + siblingDistance;
+                node.mod = node.preX - midPoint;
+            } else {
+                node.preX = midPoint;
+            }
+        }
+    },
+    // set the final xPosition of the node as its preX value and sum of
+    // all if it's parents' modifiers
+    secondWalk: function (
+        this: Highcharts.TreegraphLayout,
+        node: TreegraphNode,
+        modSum: number
+    ): void {
+        const treeLayout = this;
+        // when the chart is not iverted we want the tree to be positioned from
+        // left to right with root node close to the chart border, this is why
+        // x and y positions are switched.
+        node.yPosition = node.preX + modSum;
+        node.xPosition = node.level * treeLayout.levelSeparation;
+        node.linksFrom.forEach(function (link): void {
+            treeLayout.secondWalk(link.toNode, modSum + node.mod);
+        });
+    },
+    // shift all children of the current node from right to left.
+    executeShifts: function (
+        this: Highcharts.TreegraphLayout,
+        node: TreegraphNode
+    ): void {
+        let shift = 0,
+            change = 0;
+        for (let i = node.linksFrom.length - 1; i >= 0; i--) {
+            let link = node.linksFrom[i];
+            let childNode = link.toNode as TreegraphNode;
+            childNode.preX += shift;
+            childNode.mod += shift;
+            change += childNode.change;
+            shift += childNode.shift + change;
+        }
+    },
+    apportion: function (
```

Comment:
Can we do something to make this method a little bit smaller? It is pretty long now.

Reply:
I would leave it as it is so that it is consistent with the paper, we based on.